mlir/test/Quantum/BufferizationTest.mlir

```diff
@@ -18,18 +18,38 @@
 // Measurements //
 //////////////////
 
-func.func @counts(%q0: !quantum.bit, %q1: !quantum.bit) -> (tensor<4xf64>, tensor<4xi64>) {
+func.func @counts(%q0: !quantum.bit, %q1: !quantum.bit) {
     %obs = quantum.compbasis %q0, %q1 : !quantum.obs
-    %samples:2 = quantum.counts %obs {shots=2} : tensor<4xf64>, tensor<4xi64>
-    func.return %samples#0, %samples#1 : tensor<4xf64>, tensor<4xi64>
+
+    // CHECK: [[eigval_alloc:%.+]] = memref.alloc() : memref<4xf64>
+    // CHECK: [[counts_alloc:%.+]] = memref.alloc() : memref<4xi64>
+    // CHECK: quantum.counts {{.*}} in([[eigval_alloc]] : memref<4xf64>, [[counts_alloc]] : memref<4xi64>)
+    %samples:2 = quantum.counts %obs : tensor<4xf64>, tensor<4xi64>
+
+    // CHECK: [[dyn_eigval_alloc:%.+]] = memref.alloc() : memref<4xf64>
+    // CHECK: [[dyn_counts_alloc:%.+]] = memref.alloc() : memref<4xi64>
+    // CHECK: quantum.counts {{.*}} in([[dyn_eigval_alloc]] : memref<4xf64>, [[dyn_counts_alloc]] : memref<4xi64>)
+    %dyn_samples:2 = quantum.counts %obs : tensor<4xf64>, tensor<4xi64>
+
+    func.return
```

Comment:
> usually the reason to return the values is so that they don't get optimized away as dead code if they are unused

This only happens if the pattern applicator has any folding, i.e. the "standard" `applyPatternsAndFoldGreedily`. The applicator for `quantum-bufferize` is [`applyPartialConversion`](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Quantum/Transforms/quantum_bufferize.cpp#L50), I don't think has any folding.